### DISCUSSION TITLE
Support GitHub package registry for docker images

### DISCUSSION THREAD

**Discussion by Bullrich** [community/community#45536] [Actions] [answered]
### Select Topic Area

Product Feedback

### Body

A very practical thing of GitHub Actions is that we can build our projects (be a Typescript project or a Go project) in a Docker image, deploy it and then, instead of rebuilding the docker image everytime we want to execute the action, we simply refer to that docker image using the following configuration in our `yml` file.
```yml
runs:
  using: 'docker'
  image: 'docker://debian:stretch-slim'
```

**But**, a negative thing here is that we need to use an *external* system to store such images (https://hub.docker.com). Could it be possible to use images stored in the [GitHub Package Registry](https://github.com/features/packages) instead?
This way we wouldn't need to rely in an external system with an external account and we could publish our images directly to the repo itself.

Could be easy as changing the protocol in the image field from `docker` to `github`.
```yml
runs:
  using: 'docker'
  image: 'github://debian:stretch-slim'
```

A requirement to make this viable is to **make the system authenticate automatically**, as the current [authentication system](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#authenticating-with-a-personal-access-token-classic) is quite complicated for such a simple task.

If the action instance is already authenticated and can fetch the image from GitHub Package Registry, then we could stop people from:
- Storing images in a third party service
- Having the github action *build* the image before executing it (and save us a lot of CI time)
- Storing the compiled code in the repository itself (which is what it says to do in the [Javascript Action guide](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action#commit-tag-and-push-your-action-to-github) and I, personally, don't find having binaries in the repository a good practice).

**Comment by airtower-luna** [comment, marked as answer]
There already is such a registry, see: https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry

It should work the same way as any other container registry that isn't Docker Hub, so your image URL would look similar to `docker://ghcr.io/username/my-action:v1`. Authentication should work with the `GITHUB_TOKEN`, if necessary, assuming you've granted access to the repository.

**Reply by Bullrich** [reply]
I just tried it! If the repo is public (as a GitHub action usually is) it works out of the box!

**Comment by jeffstoner** [comment]
I have a similar situation in which I've uploaded an image to the GHES Container Registry. I would like to __not__ mark the image as public, instead, keeping it _internal_ (to the Enterprise.) Is this possible?
